### DISCUSSION TITLE
Equality connect constrain for a UPS- Parallel robot act unexpectedly

### DISCUSSION THREAD

**Discussion by AmirSamanMirjalili** [google-deepmind/mujoco#2294] [Asking for Help] [answered]
### Intro

Hi!

I am a undergrad student , I use MuJoCo for my research on Hyper Redundant parallel robots that has a stack of following 3UPS (universal-prismatic-spherical) 
![{11D67014-81F2-41D9-8BEB-A32BBB8A6839}](https://github.com/user-attachments/assets/c790711f-995c-4038-8372-59dae2763eb9)
.

### My setup

<mujoco model="3leg_parallel_robot_v2">
    <compiler angle="radian" />
    
    <default>
        <joint limited="true" />
        <motor ctrllimited="true" ctrlrange="-100 100" />
    </default>

    <asset>
        <material name="platform_base" rgba="0.2 0.3 0.4 1" />
        <material name="platform_top" rgba="0.4 0.5 0.6 1" />
        <material name="leg_outer" rgba="0.7 0.7 0.7 1" />
        <material name="leg_inner" rgba="0.6 0.6 0.6 1" />
        <material name="joint" rgba="0.5 0.1 0.1 1" />
        <material name="ground" rgba="0.5 0.5 0.5 1" specular="0.3" shininess="0.8"/>
    </asset>

    <visual>
        <global offwidth="1920" offheight="1080" />
        <quality shadowsize="4096" />
        <map shadowscale="0.5" />
    </visual>

    <worldbody>
        
        <light name="spotlight" pos="0 0 3" dir="0 0 -1" castshadow="true"/>
        <light name="frontlight" pos="2 2 2" dir="-1 -1 -1" castshadow="false"/>
        
        
        <geom type="plane" material="ground" size="2 2 0.1" pos="0 0 -0.1"/>
        
        
        <body name="base" pos="0 0 0">
            <geom type="cylinder" size="0.3 0.02" material="platform_base" />
            <geom type="cylinder" size="0.31 0.01" pos="0 0 0.015" material="platform_base" />
            
            
            <body name="p1_universal" pos="0.25 0 0">
                <joint name="p1_ux" type="hinge" axis="1 0 0" range="-45 45"/>
                <joint name="p1_uy" type="hinge" axis="0 1 0" range="-45 45"/>
                <geom type="sphere" size="0.02" material="joint" contype="0" conaffinity="0"/>
                
                
                <body name="leg1_outer">
                    <geom type="cylinder" fromto="0 0 0 0 0 0.3" size="0.015" material="leg_outer"/>
                    
                    
                    <body name="leg1_inner">
                        <joint name="p1_prismatic" type="slide" axis="0 0 1" range="0 0.3"/>
                        <geom type="cylinder" fromto="0 0 0 0 0 0.3" size="0.01" material="leg_inner"/>
                        <site name="leg1_top" pos="0 0 0.3" size="0.02" rgba="0.5 0.1 0.1 0.5"/>
                    </body>
                </body>
            </body>

            
            <body name="p2_universal" pos="-0.125 0.217 0">
                <joint name="p2_ux" type="hinge" axis="1 0 0" range="-45 45"/>
                <joint name="p2_uy" type="hinge" axis="0 1 0" range="-45 45"/>
                <geom type="sphere" size="0.02" material="joint" contype="0" conaffinity="0"/>
                
                
                <body name="leg2_outer">
                    <geom type="cylinder" fromto="0 0 0 0 0 0.3" size="0.015" material="leg_outer"/>
                    
                    
                    <body name="leg2_inner">
                        <joint name="p2_prismatic" type="slide" axis="0 0 1" range="0 0.3"/>
                        <geom type="cylinder" fromto="0 0 0 0 0 0.3" size="0.01" material="leg_inner"/>
                        <site name="leg2_top" pos="0 0 0.3" size="0.02" rgba="0.5 0.1 0.1 0.5"/>
                    </body>
                </body>
            </body>

            
            <body name="p3_universal" pos="-0.125 -0.217 0">
                <joint name="p3_ux" type="hinge" axis="1 0 0" range="-45 45"/>
                <joint name="p3_uy" type="hinge" axis="0 1 0" range="-45 45"/>
                <geom type="sphere" size="0.02" material="joint" contype="0" conaffinity="0"/>
                
                
                <body name="leg3_outer">
                    <geom type="cylinder" fromto="0 0 0 0 0 0.3" size="0.015" material="leg_outer"/>
                    
                    
                    <body name="leg3_inner">
                        <joint name="p3_prismatic" type="slide" axis="0 0 1" range="0 0.3"/>
                        <geom type="cylinder" fromto="0 0 0 0 0 0.3" size="0.01" material="leg_inner"/>
                        <site name="leg3_top" pos="0 0 0.3" size="0.02" rgba="0.5 0.1 0.1 0.5"/>
                    </body>
                </body>
            </body>
        </body>

        
        <body name="top_platform" pos="0 0 0.5">
            <joint type="free"/>
            <geom type="cylinder" size="0.3 0.02" material="platform_top" />
            <geom type="cylinder" size="0.31 0.01" pos="0 0 -0.015" material="platform_top" />
            <site name="platform_attach1" pos="0.25 0 0" size="0.02" rgba="0.5 0.1 0.1 0.5"/>
            <site name="platform_attach2" pos="-0.125 0.217 0" size="0.02" rgba="0.5 0.1 0.1 0.5"/>
            <site name="platform_attach3" pos="-0.125 -0.217 0" size="0.02" rgba="0.5 0.1 0.1 0.5"/>
        </body>
    </worldbody>

    <equality>
        
        <connect body1="leg1_inner" body2="top_platform" anchor="0.25 0 0.5"/>
        <connect body1="leg2_inner" body2="top_platform" anchor="-0.125 0.217 0.5"/>
        <connect body1="leg3_inner" body2="top_platform" anchor="-0.125 -0.217 0.5"/>
    </equality>

    <actuator>
        
        <motor joint="p1_prismatic" name="leg1_motor" gear="1"/>
        <motor joint="p2_prismatic" name="leg2_motor" gear="1"/>
        <motor joint="p3_prismatic" name="leg3_motor" gear="1"/>
    </actuator>
</mujoco>

### My question

The problem that I encountered is shown in the following image. my spherical joints do not attach to my moving platform. do you have any suggestions on how I can solve this? in general, if I assume that the inertia of my legs can be neglected what do you suggest to use in order to mimic the prismatic effect accurately and I can command the length of the joint (this is the second priority for now)
![{BB37F1D0-CFFF-42B7-BD09-17F94DBB3A59}](https://github.com/user-attachments/assets/86b90e57-6989-46c9-affa-7f5556d26c79)

### Confirmations

- [X] I searched the [latest documentation](https://mujoco.readthedocs.io/en/latest/overview.html) thoroughly before posting.
- [X] I searched previous [Issues](https://github.com/google-deepmind/mujoco/issues) and [Discussions](https://github.com/google-deepmind/mujoco/discussions), I am certain this has not been raised before.

**Comment by yuvaltassa** [comment, marked as answer]
Try using the site-based connect definition. [example](https://github.com/google-deepmind/mujoco/blob/main/test/engine/testdata/equality_site.xml)

**Reply by AmirSamanMirjalili** [reply]
Thank you @yuvaltassa. I managed to fix this using your suggestion, but I have another question: I want to know how accurate the calculated Jacobian (mj_jac) is for these types of robots (which have closed chains). Can I rely on them? or should I use a different approach

**Reply by yuvaltassa** [reply]
You cannot, they are incorrect. They don't take the constraint into account.

**Reply by AmirSamanMirjalili** [reply]
@yuvaltassa Thank you very much for your help
